### PR TITLE
Blog post: How ScopeLift Built a Flex Voting aToken on Aave

### PR DESCRIPTION
**Pull request by davidlaprade**

### CODE REVIEW

File: _posts/2022-12-16-aave-rebasing.md [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+If you deposited [DAI](https://polygonscan.com/address/0x8f3cf7ad23cd3cadbd9735aff958023239c6a063) on Polygon, you would receive [aPolDAI](https://polygonscan.com/address/0x82E64f49Ed5EC1bC6e43DAD4FC8Af9bb3A2312EE) back.
+[Here](https://docs.aave.com/developers/deployed-contracts/v3-mainnet/polygon#tokens) is a list of other commonly-issued aTokens.
+
+aTokens are _receipt tokens_: they represent a claim on deposited assets.
+When you want to [withdraw](https://github.com/aave/aave-v3-core/blob/f3e037b3638e3b7c98f0c09c56c5efde54f7c5d2/contracts/protocol/pool/Pool.sol#L197-L217) your deposit and claim its accumulated interest, your aTokens are [burned](https://github.com/aave/aave-v3-core/blob/f3e037b3638e3b7c98f0c09c56c5efde54f7c5d2/contracts/protocol/libraries/logic/SupplyLogic.sol#L131-L136).
+
+One thing that makes Aave's aTokens interesting is that their balance programmatically increases over time.
+If you have (say) 100
+[aPolWETH](https://polygonscan.com/address/0xe50fA9b3c56FfB159cB0FCA61F5c9D750e8128c8) today, you will have 102 aPolWETH in a year from now.
```
Comment (mds1) [inline _posts/2022-12-16-aave-rebasing.md]:
```suggestion
[aPolWETH](https://polygonscan.com/address/0xe50fA9b3c56FfB159cB0FCA61F5c9D750e8128c8) today, at an interest rate of 2% APY, you will have 102 aPolWETH in a year from now.
```

File: _posts/2022-12-16-aave-rebasing.md [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+  reserveCache.currLiquidityIndex
+);
+{% endhighlight %}
+
+For example, if the current liquidity index is `1050000000000000000000000000`
+(i.e. 5%) and another 2.5% has accrued since the last update, the new index will
+be `1076250000000000000000000000` (because 1.05 * 1.025 = 1.07625).
+
+It might not be obvious, but an increase in the liquidity index _just is_ the accrual of interest within Aave.
+The former constitutes the latter.
```
Comment (mds1) [inline _posts/2022-12-16-aave-rebasing.md:133]: Nit, but these two sentence were a bit confusing. Google says "constitutes" means "be a part of a whole" but really the two are identical?
```suggestion
It might not be obvious, but an increase in the liquidity index is identical to the accrual of interest within Aave.
```

File: _posts/2022-12-16-aave-rebasing.md [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+This is because any increase to the liquidity index results in a corresponding increase in aToken balance, as we'll see in the next section.
+
+#### Scaled Balances
+
+The second key thing to understand about Aave is that **it scales user balances down by the liquidity index before writing them to storage**.
+
+When you deposit (say) $100 DAI into Aave, your balance is not incremented by
+100e18 on disk. Rather, your balance is incremented by your deposit [_divided
```
Comment (mds1) [inline _posts/2022-12-16-aave-rebasing.md]: both are accurate, but "in storage" is more conventional I think
```suggestion
100e18 in storage. Rather, your balance is incremented by your deposit [_divided
```

File: _posts/2022-12-16-aave-rebasing.md [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+In other words, __the liquidity index for a certain asset is the total amount of interest earned by
+deposits of that asset into Aave__ -- expressed as a percentage scaled up by a
+[ray](https://github.com/aave/aave-v3-core/blob/f3e037b3638e3b7c98f0c09c56c5efde54f7c5d2/contracts/protocol/libraries/math/WadRayMath.sol#L17).
+
+Let's unpack that for a second, because it's a mouthful and this is really important.
+
+aPolyDAI currently has a liquidity index of `1010832786421347125844242759`, or
+1.083%. This means that _in aggregate_, over the entire lifetime of Aave, DAI deposits have earned 1.083%.
+So if you were the very first person to deposit DAI into
+Aave and you haven't touched it since, your deposit would today be worth 1.083% more.
```
Comment (mds1) [inline _posts/2022-12-16-aave-rebasing.md]: Haven't they earned 0.083%? Because `1e27 * (1 + 0.083) = 1.083e27`

```suggestion
aPolyDAI currently has a liquidity index of `1010832786421347125844242759`, or
1.083. This means that _in aggregate_, over the entire lifetime of Aave, DAI deposits have earned 1.083 - 1 = 0.083%.
So if you were the very first person to deposit DAI into
Aave and you haven't touched it since, your deposit would today be worth 0.083% more.
```

IIUC the index itself isn't a percentage, it's `1 + interest`, which becomes `1e27 + interest` as a ray, so there's a few other places below where we'd want to drop the percentage sign

File: _posts/2022-12-16-aave-rebasing.md [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+
+And that's it! That is how interest works in Aave.
+
+### Flexible Voting
+
+[Flexible voting](https://www.scopelift.co/blog/introducing-flexible-voting) (or _flex voting_ for short)
+is a novel extension to Compound-style governance systems.
+It allows delegates to split their voting weight across For/Against/Abstain options, rather than having to put all of it behind one.
+
+If flex voting could be added to aTokens, this would unlock a large amount of
```
Comment (mds1) [inline _posts/2022-12-16-aave-rebasing.md]: Should this be phrased more definitively?
```suggestion
When flex voting is to aTokens, it will unlock a large amount of
```

File: _posts/2022-12-16-aave-rebasing.md [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+
+So the main value of UNI comes from its use in voting.
+But it would be really nice if you could also deposit your UNI to earn yield at the same time.
+
+The trouble is that if you were to deposit UNI into a DeFi protocol (like Aave) you would lose your voting power.
+This is because the address that holds the tokens controls their voting weight.
+And if you deposit your UNI, the receiving contract -- not you -- would hold the tokens.
+
+So governance token holders are forced to choose: either participate in governance or earn yield.
+They cannot currently have both.
```
Comment (mds1) [inline _posts/2022-12-16-aave-rebasing.md]: They can if they deposit into element finance and use a voting vault. I don't know much about those, but maybe soften this sentence a bit as a result?
```suggestion
They cannot currently have both.
```

File: _posts/2022-12-16-aave-rebasing.md [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+
+If flex voting were added to aTokens and the Aave DAO, however, they could.
```
Comment (mds1) [inline _posts/2022-12-16-aave-rebasing.md]: Similar "if" to "when" comment

File: _posts/2022-12-16-aave-rebasing.md [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+
+In such a world, aToken holders would express their desired voting preference to
+the aToken contract, which would then cast its vote in the appropriate ratios to
+the governance contract.
+
+So, for example, if 35% of aUNI holders expressed a "No" voting preference on a
+given proposal, 60% expressed "Yes", and 5% "Abstain", then the aToken contract
+would cast its vote to the UNI governance system with 35% of its weight as
+Against, 60% of its weight as For, and 5% as Abstain.
```
Comment (mds1) [inline _posts/2022-12-16-aave-rebasing.md]: This makes it sound like all voting weight is cast, even if not all aUNI holders vote, which I don't think is true

File: _posts/2022-12-16-aave-rebasing.md [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+* a year goes by and the pool earns another 5% interest, so the new liquidity index is 1.05 * 1.05 == 1.1025
+* I now deposit $100 of the asset
+* once again, the aToken scales down my deposit and stores the result: 100 / 1.1025 == 90.70 is my balance in storage
+* if I immediately withdrew, my rebased balance would be 90.70 * 1.1025 == $100, i.e. exactly what I put in
+* Ben's rebased balance at this point would be 95.23 * 1.1025 == $105
+* so he'd have 5% more underlying than me, as expected (since the pool as a whole earned 5% while he had his money in it)
+
+And that's it! That is how interest works in Aave.
+
+### Flexible Voting
```
Comment (mds1) [inline _posts/2022-12-16-aave-rebasing.md]: I think this high-level section should be first, then the technical details after?

File: _posts/2022-12-16-aave-rebasing.md [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+
+After all, if I could have burned my aTokens and withdrawn a certain weight of
+voting tokens at the time of a proposal, any vote cast on my behalf by the aToken
+should reflect that weight.
+
+Let's look at a concrete example:
+
+* Ben deposits 100 UNI into Aave
+* a year passes and aUNI yields 50% interest (it was a _very_ good year)
+* I deposit 100 UNI
```
Comment (mds1) [inline _posts/2022-12-16-aave-rebasing.md]: Nit, maybe pick another name here and avoid "I" / first-person throughout? I don't think we've used "I" in the posts before. But we have used "we" and "our", e.g. in regards "our umbra supporters"

File: _posts/2022-12-16-aave-rebasing.md [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+As already mentioned, aToken interest rates are constantly changing
+based on supply and demand within the protocol.
+If supply goes up, interest goes down.
+If demand goes up, interest goes up.
+
+And each time the interest rate changes, the interest accrued since the last update
+is added to the liquidity index (as shown above).
+This means that the aToken liquidity index is constantly changing too.
+
+Recall that the aToken's rebased balance is just the balance on disk multiplied by the liquidity index.
```
Comment (mds1) [inline _posts/2022-12-16-aave-rebasing.md]: storage vs disk again (and right below this also)

File: _posts/2022-12-16-aave-rebasing.md [hunk 12]
```diff
@@ -0,0 +1,332 @@
...
+how can we hope to reliably compute a value based on them?
+
+Short of checkpointing both values _each time they change_ -- which would be
+unfeasibly expensive to write and later search for a popular asset with a lot of
+usage on Aave -- there seems to be no simple way to accurately compute an
+arbitrary user's aToken rebased balance at a block in the past.
+
+Fortunately, we don't have to.
+
+What we realized is that if we simply checkpoint each user's _scaled down balance_
```
Comment (mds1) [inline _posts/2022-12-16-aave-rebasing.md]:
```suggestion
Instead, if we simply checkpoint each user's _scaled down balance_
```

File: _posts/2022-12-16-aave-rebasing.md [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+`B x I` then the ratio of userA's balance to userB's is just `A / B` -- the index
+`I` cancels out.
+
+But if we only compare the stored values, wouldn't Ben and I end up with
+the same voting weight since we deposited the same amount?
+
+No.
+
+This is because Aave divides incoming balances by the current liquidity index before
+storing them.
```
Comment (mds1) [inline _posts/2022-12-16-aave-rebasing.md:283]: Feels like a lot of this section repeats info from the earlier rebasing section. IMO these two sections can be merged since voting weight is the same as user balances

**Review by mds1 — COMMENTED** [review]
Great explanations, very clear 👌

File: _posts/2022-12-16-aave-rebasing.md (hunk 4 above)
Comment (davidlaprade) [reply _posts/2022-12-16-aave-rebasing.md]: I think you might be missing a couple digits there.

> the index itself isn't a percentage, it's 1 + interest

Yup, that's right. But the index is `1010832786421347125844242759`, i.e. `1.01083`, if you subtract `1` you end up with `0.01083`, or `1.083%`

File: _posts/2022-12-16-aave-rebasing.md (hunk 2 above)
Comment (davidlaprade) [reply _posts/2022-12-16-aave-rebasing.md:133]: Huh, that's surprising.

I was using what I believe is it's [main definition](https://www.merriam-webster.com/dictionary/constitute), namely "makes up". Other ways of saying the same thing: accrual of interest is _reducible to_, is _nothing over and above_, occurs _solely in virtue of_ increases to the liquidity index.

Do you find any of those other idioms less confusing?

File: _posts/2022-12-16-aave-rebasing.md (hunk 8 above)
Comment (davidlaprade) [reply _posts/2022-12-16-aave-rebasing.md]: Fair. You're right, not all weight is cast if some aUNI holders don't express a preference

File: _posts/2022-12-16-aave-rebasing.md (hunk 10 above)
Comment (davidlaprade) [reply _posts/2022-12-16-aave-rebasing.md]: Sure, how about "Matt"? :)

File: _posts/2022-12-16-aave-rebasing.md (hunk 8 above)
Comment (davidlaprade) [reply _posts/2022-12-16-aave-rebasing.md]: > This makes it sound like all voting weight is cast even if not all aUNI holders vote

Wait, but in the example all aUNI holders express their votes. So I think this example is correct

File: _posts/2022-12-16-aave-rebasing.md (hunk 13 above)
Comment (davidlaprade) [reply _posts/2022-12-16-aave-rebasing.md:283]: Yeah, I know what you mean. Let me see what I can do

File: _posts/2022-12-16-aave-rebasing.md (hunk 2 above)
Comment (mds1) [reply _posts/2022-12-16-aave-rebasing.md:133]: Haha, to clarify it was a comment about how I thought "liquidity index == interest accrual" and not "liquidity index is one portion of interest accrual". Because the liquidity index tells you exactly how much interest accrued. So if "constitutes" is correct and "liquidity index == interest accrual" is incorrect, what are the other parts that make up interest accrual?

File: _posts/2022-12-16-aave-rebasing.md (hunk 4 above)
Comment (mds1) [reply _posts/2022-12-16-aave-rebasing.md]: Ah, yes I clearly missed a few digits somehow 😅

Comment (mds1) [reply _posts/2022-12-16-aave-rebasing.md]: I'd mark this resolved, but seems I don't have permission to do that here

File: _posts/2022-12-16-aave-rebasing.md (hunk 8 above)
Comment (mds1) [reply _posts/2022-12-16-aave-rebasing.md]: Got it, if that's the case in this example we can leave as-is, but we should make sure it's clear somewhere that not all weight is cast if some aUNI holders don't express a preference. (Maybe that's already mentioned somewhere, not sure offhand)

File: _posts/2022-12-16-aave-rebasing.md (hunk 10 above)
Comment (mds1) [reply _posts/2022-12-16-aave-rebasing.md]: Heh fine with me. Or something like charlie and dennis since we all have some relation to philly haha 😛

File: _posts/2022-12-16-aave-rebasing.md (hunk 2 above)
Comment (davidlaprade) [reply _posts/2022-12-16-aave-rebasing.md:133]: Sorry, I don't think I was clear. There are no other parts. "Be part of a whole" isn't the meaning of "constitutes" I was using. Google gave you a secondary meaning of "constitutes". The primary meaning isn't related to parts/wholes.

The primary meaning of "X constitutes Y" is the same as:
* Y is nothing over and above X
* Y is reducible to X
* Y exists solely in virtue of X

Examples:
* signing your name to a contract constitutes agreeing to it
* getting someone into checkmate constitutes winning a game of chess
* giving someone a pink slip constitutes laying them off

Comment (mds1) [reply _posts/2022-12-16-aave-rebasing.md:133]: > There are no other parts. "Be part of a whole" isn't the meaning of "constitutes" I was using. Google gave you a secondary meaning of "constitutes". The primary meaning isn't related to parts/wholes.

Oh that's a confusing set of definitions haha. Ok sounds good to me, consider this resolved then

File: _posts/2022-12-16-aave-rebasing.md (hunk 13 above)
Comment (davidlaprade) [reply _posts/2022-12-16-aave-rebasing.md:283]: Tried to do this in 8468901

File: _posts/2022-12-16-aave-rebasing.md (hunk 9 above)
Comment (davidlaprade) [reply _posts/2022-12-16-aave-rebasing.md]: I think that's reasonable. Done in 8468901

File: _posts/2022-12-16-aave-rebasing.md (hunk 11 above)
Comment (davidlaprade) [reply _posts/2022-12-16-aave-rebasing.md]: Thanks, done in ed41538

File: _posts/2022-12-16-aave-rebasing.md (hunk 6 above)
Comment (davidlaprade) [reply _posts/2022-12-16-aave-rebasing.md]: I softened this language a bit in 8468901

File: _posts/2022-12-16-aave-rebasing.md (hunk 5 above)
Comment (davidlaprade) [reply _posts/2022-12-16-aave-rebasing.md]: Curious what others think about this. I think I prefer it phrased as a hypothetical leading up to the solution

Comment (mds1) [reply _posts/2022-12-16-aave-rebasing.md]: Cool, will defer to everyone else here. Either is fine with me

File: _posts/2022-12-16-aave-rebasing.md [hunk 14]
```diff
@@ -0,0 +1,313 @@
...
+
+While flex voting makes this dream scenario possible, it's not without technical hurdles.
+The next section will get into what these are and how we solved them.
+
+### aToken Rebasing ELI5
+
+When users deposit tokens into Aave, they receive [ERC20](https://ethereum.org/en/developers/docs/standards/tokens/erc-20/) tokens in return.
+Aave's ERC20's are called [aTokens](https://github.com/aave/aave-v3-core/blob/f3e037b3638e3b7c98f0c09c56c5efde54f7c5d2/contracts/protocol/tokenization/AToken.sol).
+If you deposited [DAI](https://polygonscan.com/address/0x8f3cf7ad23cd3cadbd9735aff958023239c6a063) on Polygon, you would receive [aPolDAI](https://polygonscan.com/address/0x82E64f49Ed5EC1bC6e43DAD4FC8Af9bb3A2312EE) back.
+[Here](https://docs.aave.com/developers/deployed-contracts/v3-mainnet/polygon#tokens) is a list of other commonly-issued aTokens.
```
Comment (mds1) [inline _posts/2022-12-16-aave-rebasing.md:58]: Nit related to the below suggestion: link to optimism tokens / use them in examples instead, since our repo forks from optimism
```suggestion
[Here](https://docs.aave.com/developers/deployed-contracts/v3-mainnet/polygon#tokens) is a list of other commonly-issued aTokens on Polygon.
```

File: _posts/2022-12-16-aave-rebasing.md (hunk 8 above)
Comment (apbendi) [reply _posts/2022-12-16-aave-rebasing.md]: We could modify the example to show some percentage of the vote not being applied and mention that. I don't feel strongly about it though. Up to @davidlaprade.
